scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1500 @@
+    .. [3] T. T. Warnock. Computational investigations of low discrepancy point
+       sets, Applications of Number Theory to Numerical
+       Analysis, Academic Press, pp. 319-343, 1972.
+
+    Examples
+    --------
+    Calculate the quality of the sample using the discrepancy:
+
+    >>> from scipy.stats import qmc
+    >>> space = np.array([[1, 3], [2, 6], [3, 2], [4, 5], [5, 1], [6, 4]])
+    >>> l_bounds = [0.5, 0.5]
+    >>> u_bounds = [6.5, 6.5]
+    >>> space = qmc.scale(space, l_bounds, u_bounds, reverse=True)
+    >>> space
+    array([[0.08333333, 0.41666667],
+           [0.25      , 0.91666667],
+           [0.41666667, 0.25      ],
+           [0.58333333, 0.75      ],
+           [0.75      , 0.08333333],
+           [0.91666667, 0.58333333]])
+    >>> qmc.discrepancy(space)
+    0.008142039609053464
+
+    """
+    sample = np.asarray(sample)
+
+    n, d = sample.shape
+
+    if iterative:
+        n += 1
+
+    if method == 'CD':
+        abs_ = abs(sample - 0.5)
+        disc1 = np.sum(np.prod(1 + 0.5 * abs_ - 0.5 * abs_ ** 2, axis=1))
+
+        prod_arr = 1
+        for i in range(d):
+            s0 = sample[:, i]
+            prod_arr *= (1 +
+                         0.5 * abs(s0[:, None] - 0.5) + 0.5 * abs(s0 - 0.5) -
+                         0.5 * abs(s0[:, None] - s0))
+        disc2 = prod_arr.sum()
+
+        return ((13.0 / 12.0) ** d - 2.0 / n * disc1 +
+                1.0 / (n ** 2) * disc2)
+    elif method == 'WD':
+        prod_arr = 1
+        for i in range(d):
+            s0 = sample[:, i]
+            x_kikj = abs(s0[:, None] - s0)
+            prod_arr *= 3.0 / 2.0 - x_kikj + x_kikj ** 2
+
+        return - (4.0 / 3.0) ** d + 1.0 / (n ** 2) * prod_arr.sum()
+    elif method == 'MD':
+        abs_ = abs(sample - 0.5)
+        disc1 = np.sum(np.prod(5.0 / 3.0 - 0.25 * abs_ - 0.25 * abs_ ** 2,
+                               axis=1))
+
+        prod_arr = 1
+        for i in range(d):
+            s0 = sample[:, i]
+            prod_arr *= (15.0 / 8.0 -
+                         0.25 * abs(s0[:, None] - 0.5) - 0.25 * abs(s0 - 0.5) -
+                         3.0 / 4.0 * abs(s0[:, None] - s0) +
+                         0.5 * abs(s0[:, None] - s0) ** 2)
+        disc2 = prod_arr.sum()
+
+        disc = (19.0 / 12.0) ** d
+        disc1 = 2.0 / n * disc1
+        disc2 = 1.0 / (n ** 2) * disc2
+
+        return disc - disc1 + disc2
+    elif method == 'star':
+        return np.sqrt(
+            3 ** (-d) - 2 ** (1 - d) / n
+            * np.sum(np.prod(1 - sample ** 2, axis=1))
+            + np.sum([
+                np.prod(1 - np.maximum(sample[k, :], sample[j, :]))
+                for k in range(n) for j in range(n)
+            ]) / n ** 2
+        )
+    else:
+        raise ValueError('{} is not a valid method. Options are '
+                         'CD, WD, MD, star.'.format(method))
+
+
+def _update_discrepancy(x_new, sample, initial_disc):
```

Comment:
`_update_discrepancy`, `_perturb_discrepancy`, `n_primes`, and `primes_from_2_to` don't appear to be public, nor are they used in any public function, so please remove them. If you want to make them public or use them in a public function later, that can be done in a separate PR.

Reply:
`n_prime` and `primes_from_2_to ` are used by `Halton`. Both `_update_discrepancy ` and `_perturb_discrepancy ` are helpful for someone who wants to perturb a LHS. You will not find `_update_discrepancy ` in the literature as I derived it myself. As for `_perturb_discrepancy `, it is known in the literature but there are major mistakes in the articles. Which makes it quite nice to have it there then.

These are advanced functions, yet useful for people playing with discrepancy, and we decided to keep them private when we started this PR a while ago now. This was to expose only the common functions.

Reply:
@rgommers do we allow private functions that are not used by any public functions and are intended to be "advanced" functions? I have been writing that these should be removed, so I thought I should confirm.

Reply:
Normally we should get rid of them, unless there's a plan to start using them in a follow-up PR. In that case I don't mind leaving them in, as long as they're clearly marked with a comment about their purpose.

Reply:
There will be used in a follow-up PR where I have the `OptimalDesign` (note that this PR is a possible use of these 2 functions, so regardless of this PR, other strategies could be factored). The code is here https://github.com/tupui/scipy/tree/qmc_optimal_design

Reply:
@rgommers, if you or someone else is willing to review them, it's fine with me to keep them. 

If not, then again, this PR is already enormous. Since `OptimalDesign` was removed for now, I ask you to remove these, as they aren't much good to anyone until they are public.  If you still disagree, can you help me understand why? Why is it worth delaying completion of the PR (and stressing your volunteer reviewer) to have them in now?

Reply:
I propose to do the following compromise: I will make `_update_discrepancy ` public and remove `_perturb_discrepancy `. `update_discrepancy ` is a useful function for somebody who wants to construct an iterative design using a discrepancy criterion. Instead of calculating from scratch the discrepancy, you just have to compute the impact of a new point. `_perturb_discrepancy ` is doing something similar but dedicated to LHS.

Quick side note about your comment on the size of the PR. Here we are proposing a comprehensive set of features and if we want the QMC community to use this and contribute, it has to be appealing. Also, this is huge for us too. I have been working on this actively for so long now. I am only doing it on my spare time (I have zero funding or endorsement from my work). And it takes me literally hours to answer you comments and rework everything. Still, I appreciate all the work you are putting into this, but you don't have to bear all the weight of the review. Both @tylerjereddy and @rgommers had a look and reviewing again and again the same part of the code takes a huge effort for everybody.

Reply:
> @rgommers, if you or someone else is willing to review them, it's fine with me to keep them.
> 
> If not, then again, this PR is already enormous. Since `OptimalDesign` was removed for now, please reconsider removing these, as they aren't much good to anyone until they are public. If you still disagree, can you help me understand why? Why is it worth delaying completion of the PR (and stressing your poor volunteer reviewer) to have them in now?

You don't have to take full responsibility for every last potential bug and typo in this piece of code. What I'm taking into account is the following:
- The code quality is high, and the code is well-documented.
- Both @tupui and @Balandat are domain experts, so for algorithmic correctness and usefulness I'm happy to rely largely on their judgement and give this only a cursory look (e.g., referenced equations in article are present and seem to match).
- There's a trade-off here between review effort and potential gain. Often it's better to go with the path that's both lower effort and more motivating for the contributor.



> Quick side note about your comment on the size of the PR. Here we are proposing a comprehensive set of features and if we want the QMC community to use this and contribute, it has to be appealing. Also, this is huge for us too.

I agree. My ideal outcome is that the community likes this, and that you are interested to keep maintaining and improving this; and that other domain experts pitch in too with the occasional bug report, feature request or PR.

@mdhaber this is the model that worked well for other contributions too, from upstream devs for the likes of PocketFFT and OpenBLAS to researchers who contribute published algorithms like for `optimize.shgo` and the recent `csgraph` improvements. Every review decision and PR merge is a judgement call, and I don't always understand every last detail of stuff I review (wish I could, but I'd either need to be much smarter or need only 2 hours of sleep a day) - which works out fine in practice.

Reply:
Thanks for your patience, @tupui.

Reply:
> Both @tupui and @Balandat are domain experts, so for algorithmic correctness and usefulness I'm happy to rely largely on their judgement and give this only a cursory look (e.g., referenced equations in article are present and seem to match).

I would add that both @ArtOwen and @fjhickernell have been involved in helping validating the correctness of the methods and implementation, and arguably you will have a hard time finding any better experts on the subject. They can also speak to the usefulness of this PR to the community

Reply:
Thanks. I didn't realize that there had been detailed review already.